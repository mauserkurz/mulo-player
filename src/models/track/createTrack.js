const DEFAULT_BLOB = null;

class Track {
  constructor({
    id = '',
    name = '',
    blob = DEFAULT_BLOB,
    lastModifiedDate = new Date().getTime(),
    isLoading = false,
  }) {
    this.id = id;
    this.name = name;
    this.blob = blob;
    this.lastModifiedDate = lastModifiedDate;
    this.isLoading = isLoading;
  }

  get isLoaded() {
    return this.blob !== DEFAULT_BLOB;
  }
}

export default options => new Track(options);
