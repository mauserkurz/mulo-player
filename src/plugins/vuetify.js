import Vue from 'vue';
import Vuetify from 'vuetify';
import { THEME as theme } from '@/constants';

Vue.use(Vuetify);

export default new Vuetify({ theme });
