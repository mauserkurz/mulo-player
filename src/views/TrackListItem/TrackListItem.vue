<template>
  <v-list-item
    :class="classMap"
    transition="slide-y-transition"
    @click="switchTrack">
    <v-list-item-icon>
      <v-icon
        v-if="!track.isLoading"
        v-text="icon"
        @click.stop="getTrack"/>

      <v-progress-circular
        v-else
        :size="24"
        indeterminate
        @click.stop="cancelGettingTrack"/>
    </v-list-item-icon>

    <v-list-item-content>
      <v-list-item-title v-text="track.name"/>
    </v-list-item-content>
  </v-list-item>
</template>

<script>
export default {
  name: 'TrackListItem',

  props: {
    track: {
      type: Object,
      required: true,
    },

    currentTrackID: {
      type: String,
      required: true,
    },
  },

  computed: {
    classMap() {
      return {
        'v-list-item--active': this.currentTrackID === this.track.id,
        'track-list-item': true,
      };
    },

    icon() {
      return this.track.isLoaded ? 'mdi-check-circle' : 'mdi-cloud-download';
    },
  },

  methods: {
    switchTrack() {
      this.$emit('switch-track', this.track.id);
    },

    getTrack() {
      this.$emit('get-track', this.track.id);
    },

    cancelGettingTrack() {
      this.$emit('cancel-getting-track', this.track.id);
    },
  },
};
</script>

<style scoped lang="less">
.track-list-item {}
</style>
