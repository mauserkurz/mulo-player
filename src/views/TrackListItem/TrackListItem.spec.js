import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import vuetify from '@/plugins/vuetify';
import createTrack from '@/models/track/createTrack';
import TrackListItem from './TrackListItem.vue';

const localVue = createLocalVue();

describe('Component TrackListItem', () => {
  const defaultProps = {
    track: createTrack({ id: 1, name: 'Band 0 - Song 0' }),
    currentTrackID: '2',
  };
  const createWrapper = ({ isWithoutStubs = false, propsData = {}, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    const wrapper = renderer(TrackListItem, {
      localVue,
      vuetify,
      propsData: Object.assign({}, defaultProps, propsData),
      ...options,
    });

    return { wrapper };
  };

  describe('snapshots', () => {
    it('should match snapshot', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should show spinner when track loading', () => {
      const track = createTrack({ id: 1, isLoading: true });
      const { wrapper } = createWrapper({ isWithoutStubs: true, propsData: { track } });

      expect(wrapper.find('.v-progress-circular').exists()).toBe(true);
    });

    it('should add v-list-item--active class for current track', () => {
      const { wrapper } = createWrapper({
        isWithoutStubs: true,
        propsData: { currentTrackID: defaultProps.track.id },
      });

      expect(wrapper.find('.v-list-item--active').exists()).toBe(true);
    });
  });

  describe('events', () => {
    it('should emit switch-track event after click on list item', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.trigger('click');
      expect(wrapper.emitted()['switch-track']).toEqual([[defaultProps.track.id]]);
    });

    it('should emit get-track event after click on download icon', () => {
      const { wrapper } = createWrapper({
        stubs: {
          'v-icon': {
            template: '<div class="v-icon" @click="$listeners.click" />',
          },
        },
      });

      wrapper.find('.v-icon').trigger('click');
      expect(wrapper.emitted()['get-track']).toEqual([[defaultProps.track.id]]);
    });

    it('should emit cancel-getting-track event after click on spinner icon', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });
      const track = createTrack({ id: 1, isLoading: true });

      wrapper.setProps({ track });
      wrapper.find('.v-progress-circular').trigger('click');
      expect(wrapper.emitted()['cancel-getting-track']).toEqual([[track.id]]);
    });
  });
});
