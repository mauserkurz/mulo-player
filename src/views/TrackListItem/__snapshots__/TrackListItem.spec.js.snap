// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component TrackListItem snapshots should match snapshot 1`] = `
<div tabindex="0" transition="slide-y-transition" class="v-list-item v-list-item--link theme--light track-list-item">
  <div class="v-list-item__icon"><button type="button" class="v-icon notranslate v-icon--link mdi mdi-cloud-download theme--light"></button></div>
  <div class="v-list-item__content">
    <div class="v-list-item__title">Band 0 - Song 0</div>
  </div>
</div>
`;
